adagrams/game.py

```diff
@@ -1,11 +1,193 @@
+    bank_copy = letter_bank.copy()
+    # loop through word
+    for letter in word.upper():
+        if letter not in bank_copy:
+            return False
+        else:
+            bank_copy.remove(letter)
+    return True
+
 
 def score_word(word):
-    pass
+    #create a dictionary with these point values
+    '''Create a dictionary score_board with key-value pairs 
+    that are letter and corresponding value.
+    Initialize the result to 0
+    Loop through the characters in the word
+    Check if the character is in the letter_vals dictionary
+    If it is, add the value of the letter to the result
+    Return the result'''
+
+    result = 0
+    extra_point = [7,8,9,10]
+    
+    values_dict = {
+        'A': 1, 
+        'B': 3, 
+        'C': 3, 
+        'D': 2, 
+        'E': 1, 
+        'F': 4, 
+        'G': 2, 
+        'H': 4, 
+        'I': 1, 
+        'J': 8, 
+        'K': 5, 
+        'L': 1, 
+        'M': 3, 
+        'N': 1, 
+        'O': 1, 
+        'P': 3, 
+        'Q': 10, 
+        'R': 1, 
+        'S': 1, 
+        'T': 1, 
+        'U': 1, 
+        'V': 4, 
+        'W': 4, 
+        'X': 8, 
+        'Y': 4, 
+        'Z': 10
+    }
+
+
+    for letter in word:
+        if letter.isalpha():
+            result += values_dict[letter.upper()]
+        else:
+            continue
+    if len(word) in extra_point:
+        result += 8
+    return result
+
+
+
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    """
+    Finds highest scoring word and returns a tuple with winning word and score
+    In case of a tie:
+    - 10 letter word wins
+    - if no 10 letter words, shortest word wins
+    - if two words of same length, whichever comes first in the word list wins
+
+    Parameters: a list of strings
+
+    Returns: a tuple with one string and an integer
+    """
+    word_score_list = []
+    highest_scoring_words = []
+    high_score = 0
+
+    # get scores of each word in list, add to new list, and look for highest score
+    for word in word_list:
+        word_score_list.append(score_word(word))
+        if score_word(word) > high_score:
+            high_score = score_word(word)
+    # if only one highest score, return tuple
+    if word_score_list.count(high_score) == 1:
+        high_score_index = word_score_list.index(high_score)
+        return (word_list[high_score_index], high_score,)
+    else:
+        # if tie by score alone, start making list of words with highest score
+        for i in range(len(word_score_list)):
+            if word_score_list[i] == high_score:
+                # if 10 letter word is found, it wins - doesn't matter if there are multiples, b/c 1st one wins
+                if len(word_list[i]) == 10:
+                    return (word_list[i], high_score,)
+                else:
+                    # create list of tie words that are not 10 characters
+                    highest_scoring_words.append(word_list[i])
```

Comment:
The current implementation works, but in the worst-case run time it requires 2 loops the size of the data set and another loop over the highest scored items. There are many ways we could approach this problem, but one way we can reduce the complexity, is if we started by making a list of all the words with the highest score:
```
high_score = 0
highest_scored = []

for word in word_list:
    word_score = score_word(word)
    if word_score > high_score:
        high_score = word_score
        highest_scored = [word]
    elif word_score == high_score:
        highest_scored.append(word)

# tie breaking code would be here or in a helper function
```
In this approach we'd need to make one pass over the full `word_list` data set to find the highest scored, and then we would make one pass over the reduced set of highest_scored words to tiebreak if necessary.